noxfile.py

```diff
@@ -0,0 +1,247 @@
+# GT4Py - GridTools Framework
+#
+# Copyright (c) 2014-2024, ETH Zurich
+# All rights reserved.
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+
+from __future__ import annotations
+
+import os
+import pathlib
+import types
+from collections.abc import Sequence
+from typing import Final, Literal, TypeAlias
+
+import nox
+
+try:
+    import pytest
+
+    NO_TESTS_COLLECTED_EXIT_CODE = pytest.ExitCode.NO_TESTS_COLLECTED
+except ImportError:
+    NO_TESTS_COLLECTED_EXIT_CODE = 5
+
+# -- nox configuration --
+nox.options.default_venv_backend = "uv"
+nox.options.sessions = [
+    "test_cartesian-3.10(internal, cpu)",
+    "test_cartesian-3.10(dace, cpu)",
+    "test_cartesian-3.11(internal, cpu)",
+    "test_cartesian-3.11(dace, cpu)",
+    "test_eve-3.10",
+    "test_eve-3.11",
+    "test_next-3.10(internal, cpu, nomesh)",
+    "test_next-3.10(dace, cpu, nomesh)",
+    "test_next-3.11(internal, cpu, nomesh)",
+    "test_next-3.11(dace, cpu, nomesh)",
+    "test_storage-3.10(cpu)",
+    "test_storage-3.11(cpu)",
+]
+
+# -- Parameter sets --
+DeviceOption: TypeAlias = Literal["cpu", "cuda11", "cuda12", "rocm4_3", "rocm5_0"]
+DeviceNoxParam: Final = types.SimpleNamespace(
+    **{device: nox.param(device, id=device, tags=[device]) for device in DeviceOption.__args__}
+)
+DeviceTestSettings: Final[dict[str, dict[str, Sequence]]] = {
+    "cpu": {"extras": [], "markers": ["not requires_gpu"]},
+    **{
+        device: {"extras": [device], "markers": ["requires_gpu"]}
+        for device in ["cuda11", "cuda12", "rocm4_3", "rocm5_0"]
+    },
+}
+
+CodeGenOption: TypeAlias = Literal["internal", "dace"]
+CodeGenNoxParam: Final = types.SimpleNamespace(
+    **{
+        codegen: nox.param(codegen, id=codegen, tags=[codegen])
+        for codegen in CodeGenOption.__args__
+    }
+)
+CodeGenTestSettings: Final[dict[str, dict[str, Sequence]]] = {
+    "internal": {"extras": [], "markers": ["not requires_dace"]},
+    "dace": {"extras": ["dace"], "markers": ["requires_dace"]},
+}
+
+
+# -- nox sessions --
+@nox.session(python=["3.10", "3.11"], tags=["cartesian"])
+@nox.parametrize("device", [DeviceNoxParam.cpu, DeviceNoxParam.cuda12])
```

Comment:
why only cuda12 for gpu backends? it's the only option that we use, but if someone would like to try rocm, it would have to be added here?

Reply:
The actual session function works for any `DeviceOption` value, including `rocm` devices. We can add any other `DeviceOption` value to the parametrization at any time, the only problem is that the full matrix of parametrization will be run if someone forgets to specify the correct device parameter in the test session, and we don't usually run the tests on machines with ROCm devices...